File: js/Samples/StringCalculator.js
 /**
  * 
  * @param {string} expr 
  * @param {string[]} delimiters 
  * @returns {number[]}
  */
 function parseNumbers(expr, delimiters) {
     return expr.split(delimiters).map(Number);
 }

 /**
 * @param {string} text The string
 */
 function parseDelimiters(text) {
     let result = {
         delimiters: "",
         processedText: text
     }
     if (text.startsWith("//") && text.length > 2) {
         result.delimiters = text[2];
         result.processedText = text.split("\n")[1];
     }
     else {
         result.delimiters = /[\n,]+/;
     }

     return result;
 }
 
 /**
  * 
  * @param {number[]} numbers 
  */
 function failOnNegatives(numbers) {
     var negatives = numbers.filter(n => n < 0);
     if (negatives.length > 0){
         throw new Error("negatives not allowed: " + negatives.join(", "));
     }
 }

export default class StringCalculator {
    /**
     * 
     * @param {string} expr 
     */
    add(expr) {
        let {delimiters, processedText} = parseDelimiters(expr);
        var parsedNumbers = parseNumbers(processedText, delimiters);

        failOnNegatives(parsedNumbers);
        return parsedNumbers.length > 0 
            ? parsedNumbers.reduce((a, b) => a + b)
            : 0;  
    };
};

File: cs/Samples/StringCalculator.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Samples
{
    public class StringCalculator
    {
        public int Add(string expr)
        {
            var delimiters = ParseDelimiters(ref expr);
            var parsedNumbers = ParseNumbers(expr, delimiters);
            FailOnNegatives(parsedNumbers);
            return parsedNumbers.Any()
                ? parsedNumbers.Sum()
                : 0;
        }

        public static List<int> ParseNumbers(string expr, char[] delimiters)
        {
            return expr.Split(delimiters, StringSplitOptions.RemoveEmptyEntries)
                .Select(int.Parse).ToList();
        }

        public static char[] ParseDelimiters(ref string expr)
        {
            if (!expr.StartsWith("//") || expr.Length <= 2)
                return new[] {',', '\n'};
            var delimiter = expr[2];
            expr = expr.Split('\n')[1];
            return new[] {delimiter};
        }

        public static void FailOnNegatives(List<int> numbers)
        {
            var negatives = numbers.Where(n => n < 0).ToList();
            if (negatives.Any())
                throw new ArgumentException(
                    "negatives not allowed: " + string.Join(", ", negatives));
        }
    }
}